shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1415,83 +1415,156 @@ module MakeImpl<InputSig Lang> {
           )
         }
 
-        bindingset[call, ctx]
-        pragma[inline_late]
-        private DataFlowCallable viableImplCallContextReducedInlineLate(
-          DataFlowCall call, CcCall ctx
+        pragma[nomagic]
+        private predicate fwdFlowIntoArg(
+          ArgNodeEx arg, FlowState state, Cc outercc, ParamNodeOption summaryCtx, TypOption argT,
+          ApOption argAp, Typ t, Ap ap, ApApprox apa, boolean cc
         ) {
-          result = viableImplCallContextReduced(call, ctx)
+          fwdFlow(arg, state, outercc, summaryCtx, argT, argAp, t, ap, apa) and
+          if outercc instanceof CcCall then cc = true else cc = false
         }
 
-        bindingset[arg, ctx]
-        pragma[inline_late]
-        private DataFlowCallable viableImplCallContextReducedInlineLate(
-          DataFlowCall call, ArgNodeEx arg, CcCall ctx
-        ) {
-          call = arg.getCall() and
-          result = viableImplCallContextReducedInlineLate(call, ctx)
-        }
+        private signature module FwdFlowInInputSig {
+          default predicate callRestriction(DataFlowCall call) { any() }
 
-        bindingset[call]
-        pragma[inline_late]
-        private predicate flowIntoCallApaInlineLate(
-          DataFlowCall call, DataFlowCallable c, ArgNodeEx arg, ParamNodeEx p,
-          boolean allowsFieldFlow, ApApprox apa
-        ) {
-          PrevStage::callEdgeArgParam(call, c, arg, p, allowsFieldFlow, apa)
+          bindingset[p, apa]
+          default predicate parameterRestriction(ParamNodeEx p, ApApprox apa) { any() }
         }
 
-        bindingset[call, ctx]
-        pragma[inline_late]
-        private predicate viableImplNotCallContextReducedInlineLate(DataFlowCall call, Cc ctx) {
-          viableImplNotCallContextReduced(call, ctx)
-        }
+        /**
+         * Exposes the inlined predicate `fwdFlowIn`, which is used to calculate both
+         * flow in and flow through.
```

Comment:
So this means that the contents of this module is calculated in both contexts where one is essentially is a subset of the other. Hmm. All for the purpose of being able to project slightly earlier. This trade-off sounds dubious.

Reply:
This is the change that matters most to DB that I used to construct the performance numbers mentioned in the PR description. I reverified this by `nomagic`ing the inlined `fwdFlowIn`, which results in
```
4m56s |   426 | 13.7s @ 223  | DataFlowImpl#248dabc3::MakeImpl#DataFlowImplSpecific#21008cd7::RubyDataFlow#::Impl#DataFlow#167ac380::DataFlowMake#DataFlowImplSpecific#21008cd7::RubyDataFlow#::Global#XSS#e59174e9::OrmTracking::Config#::C#::MkStage#Stage2#::Stage#Stage3Param#::FwdFlowIn#FwdFlowInNoRestriction#::fwdFlowIn#13#fffffffffffff@ae640x39
```